TESTS/mbed_hal/i2c/main.cpp

```diff
@@ -0,0 +1,351 @@
+    i2c_free(&obj);
+#endif
+
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_MASTER);
+    i2c_free(&obj);
+
+#if DEVICE_I2CSLAVE
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_SLAVE);
+    i2c_free(&obj);
+#endif
+
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_MASTER);
+    i2c_free(&obj);
+
+#if DEVICE_I2CSLAVE
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_SLAVE);
+    i2c_free(&obj);
+
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_SLAVE);
+    i2c_free(&obj);
+
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_SLAVE);
+    i2c_free(&obj);
+#endif
+}
+
+//  Test that i2c_frequency successfully sets given frequency
+void test_i2c_frequency()
+{
+    i2c_t obj = {};
+    uint32_t set_frequency;
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_MASTER);
+
+    i2c_capabilities_t cap;
+    i2c_get_capabilities(&cap);
+
+    set_frequency = i2c_frequency(&obj, cap.minimum_frequency);
+    TEST_ASSERT_EQUAL(cap.minimum_frequency, set_frequency);
+
+    set_frequency = i2c_frequency(&obj, (cap.minimum_frequency + cap.maximum_frequency) / 2);
+    TEST_ASSERT(set_frequency >= cap.minimum_frequency && set_frequency <= cap.maximum_frequency);
+
+    set_frequency = i2c_frequency(&obj, cap.maximum_frequency);
+    TEST_ASSERT_EQUAL(cap.maximum_frequency, set_frequency);
+    i2c_free(&obj);
+
+
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_SLAVE);
+    i2c_get_capabilities(&cap);
+    set_frequency = i2c_frequency(&obj, cap.minimum_frequency);
+
+    TEST_ASSERT_EQUAL(cap.minimum_frequency, set_frequency);
+    set_frequency = i2c_frequency(&obj, (cap.minimum_frequency + cap.maximum_frequency) / 2);
+
+    TEST_ASSERT(set_frequency >= cap.minimum_frequency && set_frequency <= cap.maximum_frequency);
+    set_frequency = i2c_frequency(&obj, cap.maximum_frequency);
+
+    TEST_ASSERT_EQUAL(cap.maximum_frequency, set_frequency);
+    i2c_free(&obj);
+}
+
+// Test that i2c_set_clock_stretching successfully enable/disable clock stretching feature
+void test_i2c_set_clock_stretching()
+{
+    i2c_t obj = {};
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_MASTER);
+
+    i2c_set_clock_stretching(&obj, true);
+    i2c_set_clock_stretching(&obj, false);
+
+    i2c_free(&obj);
+}
+
+// Test that i2c_start successfully sends start signal to the bus
+void test_i2c_start()
+{
+    i2c_t obj = {};
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_MASTER);
+
+    i2c_start(&obj);
+
+    i2c_free(&obj);
+}
+
+// Test that i2c_stop successfully sends stop signal to the bus
+void test_i2c_stop()
+{
+    i2c_t obj = {};
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_MASTER);
+
+    i2c_stop(&obj);
+
+    i2c_free(&obj);
+}
+
+// Test i2c_read/i2c_write calls for non-existent slave address
+void test_i2c_read_write()
+{
+    int ret;
+    const uint32_t size = 3;
+    uint8_t data[size];
+
+    i2c_t obj = {};
+    i2c_init(&obj, I2C_SDA, I2C_SCL, DEVICE_AS_MASTER);
+
+    ret = i2c_write(&obj, MAKE_7BIT_WRITE_ADDRESS(I2C_7BIT_ADDRESS), (void *)data, size, false);
```

Comment:
@scartmell-arm What should be return value here? (assuming there is no slave connected)
`NUCLEO_F070RB` implementation return `0`
`k64f` implementation return `I2C_ERROR_NO_SLAVE`

Reply:
It should be `I2C_NO_SLAVE`